### ISSUE TITLE
ec2 volume creation breaks with older boto in v2

### ISSUE THREAD

**Issue by laurentgo** [ansible/ansible#19724]
_From @amenonsen on July 22, 2015 5:25_

Issue type: Bug report
Ansible version: devel
Ansible configuration: NA
Environment: boto 2.25
Summary:

Creating an EC2 instance using the task below fails with a traceback: "TypeError: **init**() got an unexpected keyword argument 'encrypted".

Support for encrypted volumes seems to have been added in boto 2.33, and I have 2.25. Can this be handled more gracefully somehow? Or make the module require the appropriate version?

Steps to reproduce:

```
# Create EC2 instances using the VPC subnets and security groups and
# access key we configured above.

- name: Set up EC2 instances
  ec2:
    exact_count: 1
    count_tag: "{{ cluster_tags }}"
    image: "{{ item.0.image }}"
    region: "{{ item.0.region }}"
    group_id: "{{ item.2.group_id }}"
    instance_type: "{{ item.0.type }}"
    vpc_subnet_id: "{{ item.1.subnets[0].id }}"
    instance_tags: "{{ cluster_tags | merge(item.0.tags) }}"
    key_name: "{{ ec2_key_name }}"
    assign_public_ip: yes
    volumes: "{{ item.0.volumes }}"
    wait: yes
  with_together:
    - instances
    - vpcs.results
    - security_groups.results
  register: ec2_instances
```

Expected results:

That the instance is created correctly (which it is, with stable-1.9; 1.9 supports volumes: but not 'encrypted'). Note that I'm not trying to create an encrypted volume.

Actual results:

```
"exception": "Traceback (most recent call last):\n  File \"/home/ams/.ansible/tmp/ansible-tmp-1437541681.83-138879933302241/ec2\", line 3174, in <module>\n    main()\n  File \"/home/ams/.ansible/tmp/ansible-tmp-1437541681.83-138879933302241/ec2\", line 1313, in main\n    (tagged_instances, instance_dict_array, new_instance_ids, changed) = enforce_count(module, ec2, vpc)\n  File \"/home/ams/.ansible/tmp/ansible-tmp-1437541681.83-138879933302241/ec2\", line 745, in enforce_count\n    = create_instances(module, ec2, vpc, override_count=to_create)\n  File \"/home/ams/.ansible/tmp/ansible-tmp-1437541681.83-138879933302241/ec2\", line 932, in create_instances\n    bdm[volume['device_name']] = create_block_device(module, ec2, volume)\n  File \"/home/ams/.ansible/tmp/ansible-tmp-1437541681.83-138879933302241/ec2\", line 707, in create_block_device\n    encrypted=volume.get('encrypted', None))\nTypeError: __init__() got an unexpected keyword argument 'encrypted'\n", 
```

_Copied from original issue: ansible/ansible-modules-core#1773_

**Comment by laurentgo** [comment]
_From @bcoca on July 22, 2015 12:26_

the code does this to prevent old versions from breaking:

```
return hasattr(boto, 'Version') and LooseVersion(boto.Version) >= LooseVersion('2.29.0')
```

can you check that this expression works on your loaded boto library?

**Comment by laurentgo** [comment]
_From @amenonsen on July 23, 2015 4:15_

I can't test directly, I'm afraid. I had the Debian python-boto 2.25 package installed, and when I understood the problem above, I did "sudo pip install -U boto" because I needed to get it working. I haven't uninstalled 2.25, but I don't know how to use it (or even if the more recent version overwrote some of the packaged files).

If someone can give me instructions on how to do it, I'm willing to test.

**Comment by laurentgo** [comment]
I have the same issue on Centos 7.2 with ansible 2.1.0.0 and python-boto 2.25.0-2.el7.

The stack trace is:

>  An exception occurred during task execution. The full traceback is:
>  Traceback (most recent call last):
>   File "/tmp/ansible__Y_tM4/ansible_module_ec2.py", line 1434, in <module>
>     main()
>   File "/tmp/ansible__Y_tM4/ansible_module_ec2.py", line 1424, in main
>     (instance_dict_array, new_instance_ids, changed) = create_instances(module, ec2, vpc)
>   File "/tmp/ansible__Y_tM4/ansible_module_ec2.py", line 1014, in create_instances
>     bdm[volume['device_name']] = create_block_device(module, ec2, volume)
>   File "/tmp/ansible__Y_tM4/ansible_module_ec2.py", line 768, in create_block_device
>     encrypted=volume.get('encrypted', None))
> TypeError: **init**() got an unexpected keyword argument 'encrypted' 

The version check mentionned in the comment from @bcoca is present in ec2_vol module, but not on ec2 module, when setting directly a device block mapping.

**Comment by laurentgo** [comment]
_From @ansibot on August 1, 2016 3:12_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by laurentgo** [comment]
_From @ansibot on September 12, 2016 21:22_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by laurentgo** [comment]
_From @ansibot on October 6, 2016 15:56_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by laurentgo** [comment]
_From @ansibot on October 22, 2016 9:24_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by laurentgo** [comment]
_From @vuninh on October 25, 2016 11:36_

i got the same issue, and i tried to upgrade boto using below command and i fixed it

#  pip install boto --upgrade

root@ansible:/etc/ansible# pip install boto --upgrade
Downloading/unpacking boto from https://pypi.python.org/packages/29/64/51e30f48b0e160d384d13aeab66d416b68e404525a34a87aded0da82f32a/boto-2.43.0-py2.py3-none-any.whl#md5=c524e27de9900cb4b58feea092992ba8
  Downloading boto-2.43.0-py2.py3-none-any.whl (1.3MB): 1.3MB downloaded
Installing collected packages: boto
  Found existing installation: boto 2.20.1
    Not uninstalling boto at /usr/lib/python2.7/dist-packages, owned by OS
Successfully installed boto
Cleaning up...

**Comment by laurentgo** [comment]
@vuninh I understand that upgrading boto will make the problem go away, but this module is supposed to detect which boto version is present and gracefully handle it. This bug is about a specific case not being handled properly.

**Comment by laurentgo** [comment]
_From @ansibot on November 10, 2016 13:4_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by laurentgo** [comment]
_From @ansibot on November 26, 2016 9:25_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by laurentgo** [comment]
_From @ansibot on December 7, 2016 18:6_

This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by ansibot** [comment]
@laurentgo Greetings! Thanks for taking the time to open this issue. In order for the community to handle your issue effectively, we need a bit more information. 

Here are the items we could not find in your description:
- component name

Please set the description of this issue with this template:
https://raw.githubusercontent.com/ansible/ansible/devel/.github/ISSUE_TEMPLATE.md

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibotdev** [comment]
@laurentgo You have not responded to information requests in this issue so we will assume it no longer affects you. If you are still interested in this, please create a new issue with the requested information.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)
